src/librustc_typeck/check/mod.rs

```diff
@@ -1311,9 +1311,40 @@ fn check_union(tcx: TyCtxt<'_>, id: hir::HirId, span: Span) {
     def.destructor(tcx); // force the destructor to be evaluated
     check_representable(tcx, span, def_id);
     check_transparent(tcx, span, def_id);
+    check_union_fields(tcx, span, def_id);
     check_packed(tcx, span, def_id);
 }
 
+/// When the `#![feature(untagged_unions)]` gate is active,
+/// check that the fields of the `union` does not contain fields that need dropping.
+fn check_union_fields(tcx: TyCtxt<'_>, _: Span, item_def_id: DefId) -> bool {
+    // Without the feature we check Copy types only later
+    if !tcx.features().untagged_unions {
+        return true;
+    }
+    let item_type = tcx.type_of(item_def_id);
+    if let ty::Adt(def, substs) = item_type.sty {
+        if def.is_union() {
```

Comment:
It looks like `check_union_fields` is only called from `check_union`, so should we not already know that this is a union?

This check looks a bit funny I have to say because it looks like it checks the type of an item, as opposed to items that are types... does that mean if we have a `static FOO: Union`, we run this check on the type of the static?

Reply:
No, this is just getting the fully generic type from the `DefId` of the definition of that type (i.e. the `union Foo {...}`), which may seem a bit silly.

Maybe it would be easier to just put this code in `check_union`?

Reply:
So `type_of` can be called on a `union` item and then returns that union?

I wouldn't call that the "type of" the item, that's why I am confused. The type of `union Foo { ... }` should be `Type` or so.

Reply:
`Type` is a kind because we don't have type universes.

`type_of(X)` merely means the `Ty<'tcx>` associated with that definition, e.g. the type that is used to look up `Y` in `X::Y`.

The `_of` suffix is a convention I came up with to make the query names flow a bit better, so `foo_of(X)` is the `foo` for that `X`, in general.

Maybe we should just drop the `_of`? But then it might not be obvious that e.g. `tcx.ty(def_id)` is querying anything.

Reply:
Although nowadays we could do `tcx.def_ty(def_id)`, `tcx.def_generics(def_id)`, etc. and it would probably work well enough.

Reply:
> Type is a kind because we don't have type universes.

Actually `i32` is a "kind", at least when I look at [this thing](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/subst/enum.UnpackedKind.html)... :/ That `Kind` type is really more like `exists k: kind, thing_of_kind k`. It certainly is not a "kind" itself, that would be an `enum Kind { Lifetime, Type, Const }`.

> The _of suffix is a convention I came up with to make the query names flow a bit better, so foo_of(X) is the foo for that X, in general.

Maybe `type_for`? `type_at`? I also like `def_ty` though.

Reply:
> Actually i32 is a "kind", at least when I look at this thing... :/

Sorry, I was referring to the type theory notion of "kind", not the `Kind` in rustc... which we should already rename to `Term` or something (idk if there is a discussion somewhere).

> Maybe `type_for`? `type_at`? I also like `def_ty` though.

Out of those three, I still think `def_foo` is the form that would work best in all cases (and the query system generally can only talk about defs, anything finer-grained than that is nested within HIR/MIR anyway).

Reply:
> that would be an `enum Kind { Lifetime, Type, Const }`.

(The `Const` here is surely indexed by a `Ty<'tcx>` as the kind of `const T: A` is not the same as `const T: B`?)

I agree we should rename to `Term` since `Kind` as in rustc is just plainly wrong.

Reply:
> (The Const here is surely indexed by a Ty<'tcx> as the kind of const T: A is not the same as const T: B?)

Well that depends on how precise our kind system is. ;)

Reply:
I'm guessing this is "resolved" or at least not relevant for the PR?

Reply:
How that? `check_union_fields` still contains that check. Maybe it should assert that this is a union? That seems to make more sense.

Reply:
Done